cmd/mt-whisper-importer-reader/conversion.go

```diff
@@ -0,0 +1,247 @@
+package main
+
+import (
+	"github.com/kisielk/whisper-go/whisper"
+	"github.com/raintank/metrictank/mdata"
+)
+
+type conversion struct {
+	archives []whisper.ArchiveInfo
+	points   map[int][]whisper.Point
+	method   string
+}
+
+func newConversion(arch []whisper.ArchiveInfo, points map[int][]whisper.Point, method string) *conversion {
+	return &conversion{archives: arch, points: points, method: method}
+}
+
+func (c *conversion) findSmallestLargestArchive(spp, nop uint32) (int, int) {
+	// find smallest archive that still contains enough data to satisfy requested range
+	largestArchiveIdx := len(c.archives) - 1
+	for i := largestArchiveIdx; i >= 0; i-- {
+		arch := c.archives[i]
+		if arch.Points*arch.SecondsPerPoint < nop*spp {
+			break
+		}
+		largestArchiveIdx = i
+	}
+
+	// find largest archive that still has a higher or equal resolution than requested
+	smallestArchiveIdx := 0
+	for i := 0; i < len(c.archives); i++ {
+		arch := c.archives[i]
+		if arch.SecondsPerPoint > spp {
+			break
+		}
+		smallestArchiveIdx = i
+	}
+
+	return smallestArchiveIdx, largestArchiveIdx
+}
+
+// generates points according to specified parameters by finding and using the best archives as input
+func (c *conversion) getPoints(retIdx int, spp, nop uint32) map[string][]whisper.Point {
+	res := make(map[string][]whisper.Point)
+
+	if len(c.points) == 0 {
+		return res
+	}
+
+	// figure out the range of archives that make sense to use for the requested specs
+	smallestArchiveIdx, largestArchiveIdx := c.findSmallestLargestArchive(spp, nop)
+	rawRes := c.archives[0].SecondsPerPoint
+
+	adjustedPoints := make(map[string]map[uint32]float64)
+	if retIdx > 0 && c.method == "avg" || c.method == "sum" {
```

Comment:
i think we should take the desired aggregations into account.
if whisper was sum, but we only desire sum, then why generate and store cnt?
in fact, this goes for everything here. for example, if whisper had max, but desired for MT is min, then why convert and store max into MT ?

Reply:
So far my approach was to import whatever data we can get out of the whisper files. If min is desired but the whisper file contains max then we simply do not have the data required to satisfy what's desired, but we do have min so let's import that.

Reply:
If the whisper file contains a summed aggregation and we can get the cnt for free which then allows us to also display avg, so then why not take it.

Reply:
perhaps best in this case to issue warnings. after all, if whisper had a rollup that is now no longer enabled, then maybe we should warn the user that they should probably enable that rollup in MT as well. otherwise they might get surprised if at first things seem to work (e.g. data imported from whisper is readable) but then later they find out that HM/MT is not configured to store an aggregation they wanted.

Reply:
Do I understand you right that we should issue a warning if the method has been set to `sum` but we additionally generate the `cnt`? Ok, can do that.
Actually at the time of the import we don't know what's enabled/disabled in the metrictank config unless we ask the user for the `storage-aggregations.conf` too. I assumed that most average users just have all aggregation methods enabled, so I figured I'll just fill as many of them as I can based on the data from whisper.

Reply:
yes we would need to look at both files. I think for HM, where aggregation methods are supposed to carefully crafted to balance a great experience vs storage cost vs $$ cost, it would make sense to warn if there's data in whisper archive that is not set up for HM. at the very least, it's an incentive to make sure the storage-schemas for HM is as desired without forgetting a needed aggregation.

Reply:
I don't think that's necessary. If a user has whisper files that contain `max` aggregations and then configures their MT to only keep `min` they'll see the imported data when they choose to `consolidateBy(max)` and the new data when they `consolidateBy(min)`, which makes sense and is probably a very rare case anyway. Asking the user to also provide their `storage-aggregations.conf` just to be able to warn them if they potentially might have misconfigured something is at average probably a worse user experience than a better one, because the vast majority is just going to have all aggregations on.

Reply:
> Asking the user to also provide their storage-aggregations.conf just to be able to warn them

we are the ones constructing this file, at least for the HM use case.

Reply:
Yeah, so then we'll need to make the user download two files and point the tool at both of them etc. I'd prefer to keep the process as simple and smooth as possible, so it doesn't seem justified to have to deal with an extra file just to display a warning